src/coreclr/src/System.Private.CoreLib/src/System/RuntimeHandles.cs

```diff
@@ -199,12 +199,103 @@ internal static bool HasElementType(RuntimeType type)
         [MethodImpl(MethodImplOptions.InternalCall)]
         internal static extern object CreateInstance(RuntimeType type, bool publicOnly, bool wrapExceptions, ref bool canBeCached, ref RuntimeMethodHandleInternal ctor, ref bool hasNoDefaultCtor);
 
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        internal static extern object Allocate(RuntimeType type);
-
         [MethodImpl(MethodImplOptions.InternalCall)]
         internal static extern object CreateInstanceForAnotherGenericParameter(RuntimeType type, RuntimeType genericParameter);
 
+        /// <summary>
+        /// Given a RuntimeType, returns information about how to activate it via calli
+        /// semantics. This method will ensure the type object is fully initialized within
+        /// the VM, but it will not call any static ctors on the type.
+        /// </summary>
+#if FEATURE_COMINTEROP
+#endif
+        internal static void GetActivationInfo(
+            RuntimeType rt,
+            bool forGetUninitializedInstance,
+            out MethodTable* pMT,
+            out delegate*<void*, object> pfnAllocator,
+            out void* vAllocatorFirstArg,
+            out delegate*<object, void> pfnCtor,
+            out bool ctorIsPublic)
+        {
+            Debug.Assert(rt != null);
+
+            // Get the requested activation information
+            // GetUninitializedInstance doesn't care about ctor information
+
+            try
+            {
+                delegate*<void*, object> pfnAllocatorTemp = default;
+                void* vAllocatorFirstArgTemp = default;
+                delegate*<object, void> pfnCtorTemp = default;
+                Interop.BOOL fCtorIsPublicTemp = default;
+                MethodTable* pMethodTableTemp = default;
+
+                _GetActivationInfo(
+                    ObjectHandleOnStack.Create(ref rt),
+                    &pfnAllocatorTemp, &vAllocatorFirstArgTemp,
+                    fUnwrapNullable: forGetUninitializedInstance ? Interop.BOOL.TRUE : Interop.BOOL.FALSE,
+                    ppfnCtor: forGetUninitializedInstance ? null : &pfnCtorTemp,
+                    pfCtorIsPublic: forGetUninitializedInstance ? null : &fCtorIsPublicTemp,
+                    &pMethodTableTemp);
```

Comment:
I was hoping that the managed side won't need to know about the MethodTable at all.

- Null can be used as sentinel value for null returning allocator:
```
static object? ReturnNull(void* _) => null;
pfnAllocator = (pfnAllocatorTemp != null) ? pfnAllocator : &NullAllocator;
```
- The PreciseCtor check can be inside the QCall. Given how much the rest costs, outlining the check is unlikely a measurable optimization.
- Asserts that are verifying that the unmanaged side have done the right thing are not really necessary.

Reply:
The biggest use case right now is to allow `ActivatorCache` to cheaply query whether the requested type is a struct type or whether it's `Nullable<T>`. The current implementations `Type.IsValueType` and `Nullable.GetUnderlyingType` both make calls back into the VM. Maybe it's ok to take that hit here since the information is being cached, but per earlier feedback I'm trying to avoid chattiness where possible.

I'm also trying to avoid building too much knowledge about `CreateInstance` and `GetUninitializedInstance` into the workhorse routine, since that will make it tricky to expand this to new scenarios in the future. (Example: `CreateFactory` would need slightly different logic.) But maybe that's a problem for the future instead of the present. 🙂

Reply:
> allow ActivatorCache to cheaply query whether the requested type is a struct type or whether it's `Nullable<T>`

The ActivatorCache needs this to finish computing the function pointers. I believe that this logic can be moved to the unmanaged side without too much trouble.

Reply:
I'm nervous making one single unmanaged method that's special-casing all possible behaviors. Now that one method needs to know about the behaviors of `Activator.CreateInstance` _and_ `GetUninitializedObject`. Assuming `Activator.CreateFactory(paramful-ctor)` comes along in the future, the method will also need to be updated to account for that.

In order to future-proof this, it really seems like this workhorse routine should be responsible for providing the information, but the callers should be responsible for interpreting the information and applying whatever custom logic they need. (Example: `GetUninitializedObject` always needs to invoke precise cctors. `Activator` never needs to invoke precise cctors, because even if it's not querying for the default ctor, it intends on calling _some_ ctor at some point.)

Reply:
I guess the argument I'm trying to make is that the unmanaged code is starting to get so much branching and special-casing inside of it, it's growing to become the combined size of both methods it's intended to replace. At which point I think combining the methods loses value, so we may as well split it back into the two methods `CreateInstance` and `GetUninitializedObject`.

Reply:
I thought that the managed side will be responsible for invoking the cctor in the `GetUninitializedObject` case. This method should not have special casing for it.

I would be fine with leaving `GetUninitializedObject` alone for this change given the number of differences between the two.